src/KeyManager.sol

```diff
@@ -33,62 +38,103 @@ abstract contract KeyManagerBase {
-    }
+    // Key derivation for encryption
 
     // Because we are using hardened derivation, for each
     // contract we will need someone to sign it off chain
-    mapping(address => bytes) public derivedPub;
+    mapping(bytes32 => bytes) internal _derivedPub;
 
     function onchain_DeriveKey(
         address a,
         bytes memory dPub,
         // Signature from a valid kettle
-        bytes memory sig
+        bytes memory sig,
+        uint32 index
     ) public {
-        bytes32 digest = keccak256(abi.encodePacked(a, dPub));
+        bytes32 digest = keccak256(abi.encodePacked(a, index, dPub));
         require(Secp256k1.verify(xPub, digest, sig));
-        derivedPub[a] = dPub;
+        bytes32 key = keccak256(abi.encodePacked(a, index));
+        _derivedPub[key] = dPub;
     }
 
-    function offchain_DeriveKey(address a) public returns (bytes memory dPub, bytes memory sig) {
-        bytes32 dPriv = _derivedPriv(a);
+    function offchain_DeriveKey(address a, uint32 index) public returns (bytes memory dPub, bytes memory sig) {
+        bytes32 dPriv = _derivedPriv(a, index);
         dPub = PKE.derivePubKey(dPriv);
-        bytes32 digest = keccak256(abi.encodePacked(a, dPub));
+        bytes32 digest = keccak256(abi.encodePacked(a, index, dPub));
         sig = Secp256k1.sign(uint256(xPriv()), digest);
         require(Secp256k1.verify(xPub, digest, sig));
     }
 }
 
 contract KeyManager_v0 is KeyManagerBase {
     IAndromeda public Suave;
+    BIP32 public bip32;
+
 
     constructor(address _Suave) {
         Suave = IAndromeda(_Suave);
+        bip32 = new BIP32(Suave);
+    }
+
+    function derivedPub(address a, uint32 index) public view returns (bytes memory) {
+        bytes32 key = keccak256(abi.encodePacked(a, index));
+        return _derivedPub[key];
+    }
+
+    function derivedPub(address a) public view returns (bytes memory) {
+        bytes32 key = keccak256(abi.encodePacked(a, uint32(0)));
+        return _derivedPub[key];
+    }
+
+    // Any contract in confidential mode can request a
+    // hardened derived key
+    function _derivedPriv(address a, uint32 index) internal override returns (bytes32) {
+        bytes32 _seed = Suave.volatileGet("seed");
+        // derive the key pair of the address first
+        uint32 addrIndex = uint32(uint256(keccak256(abi.encodePacked(a)))) | HARDENED_START_INDEX;
+        (BIP32.ExtendedPrivateKey memory _xPriv, BIP32.ExtendedPublicKey memory _xPub) = bip32.deriveChildKeyPairFromSeed(abi.encodePacked(_seed), addrIndex);
+        // if an index is provided, derive the child key pair of the given address at that index
+        if(index >= HARDENED_START_INDEX) {
+            (BIP32.ExtendedPrivateKey memory _cxPriv, BIP32.ExtendedPublicKey memory _cxPub) = bip32.deriveChildKeyPair(_xPriv, index);
+            return _cxPriv.key;
+        }
+        // if no index is provided, return the address's private key
+        return _xPriv.key;
```

Comment:
Hm, why would we not derive non-hardened?
This is surprising as we would be ignoring the index passed in. If you want to enforce hardened, maybe just `index | HARDENED_START_INDEX`?
 In the examples you are actually using `0` for indexes, which would be no different from `1` which I'd consider wrong

Reply:
Yes, our current version of the key manager is only allowing for hardened key derivation thats why we don't accept indexes below 2³¹.

Unfortunately, this is kind of dirty workaround for the first layer of the key derivation tree.
If you derive a key only for an address without passing an index, then we simply ignore the index parameter because we don't have "optional" parameter in solidity.
So that's why, when no index is being passed, we store the public key of "first layer" key derivation for the given address with the index 0. This is due to the tuple mapping above. mapping((address, index) -> bytes) derivedPub
If we want to allow non-hardened key derivation, then we would need another mapping to that stores address->bytes in addition to the above tuple mapping.